src/TwitchLib.Communication/Clients/ClientBase.cs

```diff
@@ -0,0 +1,369 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using TwitchLib.Communication.Events;
+using TwitchLib.Communication.Extensions;
+using TwitchLib.Communication.Interfaces;
+using TwitchLib.Communication.Models;
+using TwitchLib.Communication.Services;
+
+namespace TwitchLib.Communication.Clients
+{
+    /// <summary>
+    ///     This <see langword="class"/> bundles almost everything that <see cref="TcpClient"/> and <see cref="WebSocketClient"/> have in common
+    ///     to be able to 
+    ///     <list>
+    ///         <item>
+    ///             pass instances of this <see langword="class"/> to <see cref="NetworkServices{T}"/> and <see cref="Services.ThrottlerService"/>
+    ///         </item>
+    ///         <item>
+    ///             and to access Methods of this instance within <see cref="NetworkServices{T}"/> and <see cref="Services.ThrottlerService"/>
+    ///         </item>
+    ///     </list>
+    /// </summary>
+    public abstract class ClientBase<T> : IClient
+        where T : IDisposable
+    {
+        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1, 1);
+        private readonly NetworkServices<T> _networkServices;
+        private CancellationTokenSource _cancellationTokenSource;
+        
+        /// <summary>
+        ///     This <see cref="_cancellationTokenSource"/> is used for <see cref="_networkServices.ListenTask"/>
+        ///     whenever a call to <see cref="_cancellationTokenSource.Cancel()"/> is made
+        /// </summary>
+        internal CancellationToken Token => _cancellationTokenSource.Token;
+        
+        internal static TimeSpan TimeOutEstablishConnection => TimeSpan.FromSeconds(15);
```

Comment:
ill assume 15 seconds is a standard practice

Reply:
Not sure if its a standard practice but its a reasonable amount imo